### PR TITLE
Add permission checks to analyzer

### PR DESCRIPTION
**Pull request by dain**
The design and implementation of this is very much a work in progress, so please let me know if you have any feedback.

This adds basic authorization checks to Presto.  During analysis of a statement we check if the identity attached to the session is authorized to perform the specified actions.  For statements that skip analysis, the checks are performed directly in the task execution.  The implementation allows for plugins to provide a system wide access control implementation or a specific access control for connector instance.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java [hunk 1]
```diff
@@ -685,6 +724,99 @@ public void renameTable(String databaseName, String tableName, String newDatabas
...
+
+    @Override
+    public Set<HivePrivilege> getDatabasePrivileges(String user, String databaseName)
+    {
+        ImmutableSet.Builder<HivePrivilege> privileges = ImmutableSet.builder();
+
+        if (isDatabaseOwner(user, databaseName)) {
+            privileges.add(OWNERSHIP);
+        }
+        // Currently the only database priviledge checked is ownership.  If that chagnes, uncomment this line
```
Comment (zhenyuy-fb) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java]: nit: chagnes -> changes.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/HiveMetastore.java [hunk 2]
```diff
@@ -48,4 +54,41 @@
...
+
+        Optional<Database> databaseMetadata = getDatabase(databaseName);
+        if (!databaseMetadata.isPresent()) {
+            return false;
+        }
+
+        Database database = databaseMetadata.get();
+
+        // a database can be owned by a user or role
+        if (database.getOwnerType() == USER && user.equals(database.getOwnerName())) {
```
Comment (zhenyuy-fb) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/HiveMetastore.java]: per offline discussion, the ownership should be checked against OWNERSHIP privilege in Hipster.

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/HiveMetastore.java [hunk 3]
```diff
@@ -48,4 +54,41 @@
...
+            return true;
+        }
+        return false;
+    }
+
+    default boolean isTableOwner(String user, String databaseName, String tableName)
+    {
+        // a table can only be owned by a user
+        Optional<Table> table = getTable(databaseName, tableName);
+        return table.isPresent() && user.equals(table.get().getOwner());
```
Comment (zhenyuy-fb) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/HiveMetastore.java]: same as above.

**Comment by cberner** [comment]
It looks like you missed "rename column". It seems like that should have permission checks too

**Comment by cberner** [comment]
Should "reset session" also be guarded? It seems like that's a form of modifying the session, and even if it's not possible today it seems possible in theory that your session properties could have been changed from the defaults by a privileged actor

File: presto-spi/src/main/java/com/facebook/presto/spi/security/ConnectorAccessControl.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.security;
+
+import com.facebook.presto.spi.SchemaTableName;
+
+public interface ConnectorAccessControl
+{
+    void checkCanCreateTable(Identity identity, SchemaTableName tableName);
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/ConnectorAccessControl.java]: Add javadoc explaining that you should throw a `AccessDeniedException`

File: presto-spi/src/main/java/com/facebook/presto/spi/security/Identity.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+import java.security.Principal;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class Identity
+{
+    private final String user;
+    // private final String role;
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/security/Identity.java]: remove this

**Comment by cberner** [comment]
Does it make sense to have a "rename table" permission? It seems like this is the union of "drop table" and "create table"

**Comment by cberner** [comment]
I don't know how the Hive metastore security works, so check with someone that those are the right calls to be using. Otherwise looks good, besides the comments I had

**Comment by dain** [comment]
@cberner you asked "Does it make sense to have a "rename table" permission? It seems like this is the union of "drop table" and "create table". 

The difference is when you drop and recreate, you reset any grants for the table.  When you rename the grants (and views on the table), should follow the rename.

**Comment by cberner** [comment]
Ah, I see

**Pull request closed without merging**